devicemodel/hw/pci/virtio/virtio_ipu.c

```diff
@@ -0,0 +1,381 @@
+/*
+ * Copyright (C) 2018 Intel Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+/*
+ * virtio ipu
+ * IPU mediator DM
+ */
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <assert.h>
+#include <pthread.h>
+
+#include "dm.h"
+#include "pci_core.h"
+#include "virtio.h"
+#include "virtio_kernel.h"
+#include "vmmapi.h"
+
+/*
+ * Size of queue was chosen experimentally in a way
+ * that it allows to do IPU ISYS MIPI camera capture without
+ * any delay for interrupt/msg send, this value should
+ * be tuned.
+ */
+#define VIRTIO_IPU_RINGSZ 1024
+
+/*
+ * IPU mediation makes use of 2 VQs.
+ * VQ0 for buffer mgmnt & 1 time configuration;
+ * VQ1 for interrupt/msg exchange
+ */
+#define VIRTIO_IPU_VQ_NUM 2
+
+const char *ipu4_vbs_dev_path = "/dev/vbs_ipu";
+
+static int virtio_ipu_debug = 0;
+#define DPRINTF(...)\
+do {\
+	if (virtio_ipu_debug)\
+		printf(__VA_ARGS__);\
+} while (0)
+
+#define WPRINTF(...) printf(__VA_ARGS__)
+
+struct virtio_ipu {
+	struct virtio_base base;
+	struct virtio_vq_info vq[VIRTIO_IPU_VQ_NUM];
+	pthread_mutex_t mtx;
+	/* VBS-K variables */
+	struct {
+		enum VBS_K_STATUS ipu_kstatus;
+		int ipu_fd;
+		struct vbs_dev_info ipu_kdev;
+		struct vbs_vqs_info ipu_kvqs;
+	} vbs_k;
+};
+
+static int virtio_ipu_k_init(struct virtio_ipu *ipu);
+static int virtio_ipu_k_start(struct virtio_ipu *ipu);
+static int virtio_ipu_k_stop(struct virtio_ipu *ipu);
+static int virtio_ipu_k_reset(struct virtio_ipu *ipu);
+static int virtio_ipu_k_dev_set(struct vbs_dev_info *ipu_kdev,
+								const char *name, int vmid,
+								int nvq, uint32_t feature,
+								uint64_t pio_start, uint64_t pio_len);
+
+static int virtio_ipu_k_vq_set(struct vbs_vqs_info *ipu_kvqs,
+								unsigned int nvq, unsigned int idx,
+								uint16_t qsize,
+								uint32_t pfn, uint16_t msix_idx,
+								uint64_t msix_addr, uint32_t msix_data);
+
+static void virtio_ipu_no_notify(void *, struct virtio_vq_info *);
+static void virtio_ipu_set_status(void *, uint64_t);
+static void virtio_ipu_reset(void *);
+
+static struct virtio_ops virtio_ipu_ops_k = {
+	"virtio_ipu",			/* our name */
+	VIRTIO_IPU_VQ_NUM,		/* we support 2 virtqueue */
+	0,						/* config reg size */
+	virtio_ipu_reset,		/* reset */
+	virtio_ipu_no_notify,	/* device-wide qnotify */
+	NULL,					/* read virtio config */
+	NULL,					/* write virtio config */
+	NULL,					/* apply negotiated features */
+	virtio_ipu_set_status,	/* called on guest set status */
+	0,						/* our capabilities */
+};
+
+static int
+virtio_ipu_k_init(struct virtio_ipu *ipu) {
```

Comment:
code style.  We usually put the bracket to new line below for function. Fix all in this file. thanks.

Reply:
Patch set added. Please review it